### PR TITLE
Make civilians wander around when idle

### PR DESCRIPTION
**Pull request by reaperrr**
Via new `IdleWander` trait.

### CODE REVIEW

**Comment by Mailaender** [comment]
Works as promised. :white_check_mark:

File: OpenRA.Mods.Common/Traits/AttackWander.cs [hunk 1]
```diff
@@ -27,7 +27,7 @@ class AttackWander : Wanders
 		public AttackWander(Actor self, AttackWanderInfo info)
 			: base(self, info)
 		{
-			attackMove = self.TraitOrDefault<AttackMove>();
+			attackMove = self.Trait<AttackMove>();
```
Comment (Mailaender) [inline OpenRA.Mods.Common/Traits/AttackWander.cs]: :+1:

**Comment by Mailaender** [comment]
Do `CHAN` and `DELPHI` in the `ra` mod also need the disablement?

**Comment by reaperrr** [comment]
> Do CHAN and DELPHI in the ra mod also need the disablement?

`DELPHI` in RA actually has it disabled.

I don't remember any RA missions with `CHAN`, and he was wandering in original TD as well.
Also, mappers can still disable the wandering in map rules, if really necessary.

File: OpenRA.Mods.Common/Traits/IdleWander.cs [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Will Move to a random location within MoveRadius when idle.",
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/IdleWander.cs]: Why does this need its own trait? This is what `Wanders` is supposed to do!

Comment (Mailaender) [inline OpenRA.Mods.Common/Traits/IdleWander.cs]: I agree that the broken by design trait https://github.com/OpenRA/OpenRA/pull/7222/files#diff-e2c08dd74a071ebf738fa2aaa9ba81dcR86 is terrible. This might be an opportunity to fix it.

**Comment by reaperrr** [comment]
> Why does this need its own trait? This is what `Wanders` is supposed to do!

Updated accordingly.

While I was at it, I also made the trait upgradable and support both `Mobile` and `Aircraft` actors.

File: OpenRA.Mods.Common/Traits/Wanders.cs [hunk 3]
```diff
@@ -28,36 +28,49 @@ public abstract class WandersInfo : ITraitInfo
...
+		public override object Create(ActorInitializer init) { return new Wanders(init.Self, this); }
 	}
 
-	public class Wanders : INotifyIdle, INotifyBecomingIdle
+	public class Wanders : UpgradableTrait<WandersInfo>, INotifyCreated, INotifyIdle, INotifyBecomingIdle
 	{
 		readonly Actor self;
 		readonly WandersInfo info;
+		Mobile mobile;
+		Aircraft aircraft;
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Wanders.cs]: You may be able to store `IResolveOrder move = self.TraitOrDefault<IMove> as IResolveOrder` to avoid the explicit mobile/aircraft.

**Comment by reaperrr** [comment]
Updated.

**Comment by Mailaender** [comment]
Even better now. :+1:

File: OpenRA.Mods.Common/OpenRA.Mods.Common.csproj [hunk 4]
```diff
@@ -277,7 +277,7 @@
     <Compile Include="Traits\Attack\AttackOmni.cs" />
     <Compile Include="Traits\Attack\AttackSuicides.cs" />
     <Compile Include="Traits\Attack\AttackTurreted.cs" />
-    <Compile Include="Traits\Attack\AttackWander.cs" />
+    <Compile Include="Traits\AttackWander.cs" />
```
Comment (Phrohdoh) [inline OpenRA.Mods.Common/OpenRA.Mods.Common.csproj]: Why was this moved?

File: OpenRA.Mods.Common/Traits/Wanders.cs [hunk 5]
```diff
@@ -28,36 +28,47 @@ public abstract class WandersInfo : ITraitInfo
...
-		public abstract object Create(ActorInitializer init);
+		public override object Create(ActorInitializer init) { return new Wanders(init.Self, this); }
 	}
 
-	public class Wanders : INotifyIdle, INotifyBecomingIdle
+	public class Wanders : UpgradableTrait<WandersInfo>, INotifyCreated, INotifyIdle, INotifyBecomingIdle
 	{
 		readonly Actor self;
 		readonly WandersInfo info;
+		IResolveOrder move;
```
Comment (Phrohdoh) [inline OpenRA.Mods.Common/Traits/Wanders.cs]: This is... odd.
Why do you need `IMove` as an `IResolveOrder`, and can you guarantee every `IMove` is an `IResolveOrder`?

File: OpenRA.Mods.Common/Traits/Wanders.cs [hunk 6]
```diff
@@ -92,7 +103,7 @@ CPos PickTargetLocation()
 
 		public virtual void DoAction(Actor self, CPos targetCell)
 		{
-			throw new NotImplementedException("Base class Wanders does not implement method DoAction!");
+			move.ResolveOrder(self, new Order("Move", self, false) { TargetLocation = targetCell });
```
Comment (Phrohdoh) [inline OpenRA.Mods.Common/Traits/Wanders.cs]: Shouldn't you issue this order instead of directly resolving it?

Comment (pchote) [inline OpenRA.Mods.Common/Traits/Wanders.cs]: This is synced code (runs on every client), so issuing an order won't work.

File: OpenRA.Mods.Common/Traits/Wanders.cs (hunk 5 above)
Comment (reaperrr) [inline OpenRA.Mods.Common/Traits/Wanders.cs]: This approach was suggested by @pchote to avoid having to explicitly look up both `Mobile` and `Aircraft`.

Comment (pchote) [inline OpenRA.Mods.Common/Traits/Wanders.cs]: See the earlier comments for context.

File: OpenRA.Mods.Common/OpenRA.Mods.Common.csproj (hunk 4 above)
Comment (reaperrr) [inline OpenRA.Mods.Common/OpenRA.Mods.Common.csproj]: Because the trait name is misleading, this is not an `AttackBase` derivative and not strictly an attack trait by itself, either.

**Comment by obrakmann** [comment]
Looks fine to me, :+1:

**Comment by obrakmann** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/b9dab0b6c8a8c9a026aa1a128ad81b2779fd3681)
